src/mlpack/methods/cf/ncf_impl.hpp

```diff
@@ -0,0 +1,430 @@
+
+    itemRates.shed_row(0);
+    itemRates.shed_row(1);
+    // List of all items.n
+    arma::vec negativeList = arma::linspace<arma::vec>(
+        0, numItems - 1, numItems);
+
+    for (size_t j = 0; j < itemRates.n_cols; j++)
+    {
+      // Remove items which have been rated.
+      arma::uvec temp = arma::find(negativeList ==  itemRates(j));
+      if(temp.n_rows != 0)
+        negativeList.shed_row(temp(0));
+    }
+
+    std::vector<double> negList;
+    // Add all negatives to a vector.
+    negList = arma::conv_to<std::vector<double>>::from(negativeList);
+    negatives.push_back(negList);
+  }
+}
+
+/**
+ * Create training instances using both positive and negative instances.
+ */
+void NCF::GetTrainingInstance(arma::mat& predictors,
+                              arma::mat& responses)
+{
+  size_t q = 0, temp;
+  arma::colvec users(dataset.n_cols * (neg+1));
+  arma::colvec resp(dataset.n_cols * (neg+1)), items(dataset.n_cols * (neg+1));
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    temp = neg;
+
+    // Rating exists.
+    users(q) = dataset(0, i) + 1;
+    items(q) = dataset(1, i) + 1;
+    resp(q) = implicit ? 1 : dataset(2, i);
+    q++;
+    // From find negatives.
+    size_t val = negatives[dataset(0, i)].size();
+
+    while (temp != 0)
+    {
+      size_t j = math::RandInt(val);
+      // Add negatives.
+      users(q) = dataset(0, i) + 1;
+      items(q) = negatives[dataset(0, i)][j] + 1;
+      resp(q) = 0;
+      q++;
+      temp--;
+    }
+  }
+  predictors = arma::join_vert(users, items);
+  responses = arma::conv_to<arma::mat>::from(resp);
+}
+
+double NCF::Evaluate(const arma::mat& parameters,
+                     const size_t begin,
+                     const size_t batchSize,
+                     const bool deterministic)
+{
+  double loss = network.Evaluate(parameters, begin, batchSize, deterministic);
+  return loss;
+}
+
+void NCF::Gradient(const arma::mat& parameters,
+                   const size_t begin,
+                   arma::mat& gradient,
+                   const size_t batchSize)
+{
+  network.Gradient(parameters, begin, gradient, batchSize);
+  GetTrainingInstance(network.Predictors(), network.Responses());
+}
+
+/**
+ * Train the model using the given optimizer.
+ */
+template<typename OptimizerType>
+void NCF::Train(OptimizerType optimizer)
+{
+  arma::mat predictors, responses;
+
+  FindNegatives();
+
+  GetTrainingInstance(predictors, responses);
+
+  network.Predictors() = std::move(predictors);
+  network.Responses() = std::move(responses);
+
+  // Train the model.
+  Timer::Start("ncf_optimization");
+  const double out = optimizer.Optimize(this->network, network.Parameters());
+  Timer::Stop("ncf_optimization");
+
+  Log::Info << "NCF::NCF(): final objective of trained model is " << out
+      << "." << std::endl;
+}
+
+/**
+ * Create a model for GMF.
+ */
+void NCF::CreateGMF()
+{
+  size_t size = (dataset.n_cols * (neg+1));
```

Comment:
About the batch size issue, the ``Subview`` layer is the only one that depends on this information right?

Reply:
`Subview` layer and `Linear` layer. Linear is not present in the current code. I'll be pushing that in a while.